### DISCUSSION TITLE
Is there a text length limit to BackLog command?

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#19064] [Q&A]
I have found in documentation that there is a limit to 30 commands in the BackLog command, but I have issued 30 commands that got cut in the middle like in the case of Template command that was more then 500 characters.
Full length of BackLog command was more then 500 characters.

**Comment by sfromis** [comment]
Max length is not fixed, but may depend on platform and details of what you are doing. On ESP32 I had no trouble with a much longer Backlog command. And on an ESP8266 I was also able to go above 500.

Anyway, doing config with very long backlog is problematic for another reason. Quite a few of the config commands will trigger a restart, which very well can be before all commands are executed. And many commands will not work until after the restart for having the feature available.

**Reply by msetina** [reply]
OK. Most of the commands were for names of relays, switches and the device.
 In my case command length was little more of 580 characters, in the reply 130 were missing. When processing the commands, for the garbled command there was following text in the log
```
Received response: 08:58:00.533 MQT: stat/small_tasmota32/RESULT = {"Command":"Unknown"}

**Reply by sfromis** [reply]
Well, still no details of what you are doing. I had no trouble with a 603-char backlog command. While the "rule" is obviously useless, it was stored full length.
`backlog rule1 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789`

**Reply by msetina** [reply]
No problem, did't think it was needed. This is the command as reply:
```
Backlog DeviceName ESP32 enota; SetOption53 1; FriendlyName1 rele1; WebButton1 rele1; FriendlyName2 rele2; WebButton2 rele2; FriendlyName3 rele3; WebButton3 rele3; FriendlyName4 rele4; WebButton4 rele4; FriendlyName5 rele5; WebButton5 rele5; FriendlyName6 rele6; WebButton6 rele6; FriendlyName7 rele7; WebButton7 rele7; FriendlyName8 rele8; WebButton8 rele8; SwitchText1 stikalo1; SwitchMode1 16; SwitchText2 stikalo2; SwitchMode2 16; SwitchText3 stikalo3; SwitchMode3 16; SwitchText4 stikalo4; SwitchMode4 16; SwitchText5 stikalo5; SwitchMode5 16; SwitchText6 stikalo6; SwitchMode6 16;

Reply on serial:
08:57:56.443 CMD: Backlog DeviceName ESP32 enota; SetOption53 1; FriendlyName1 rele1; WebButton1 rele1; FriendlyName2 rele2; WebButton2 rele2; FriendlyName3 rele3; WebButton3 rele3; FriendlyName4 rele4; WebButton4 rele4; FriendlyName5 rele5; WebButton5 rele5; FriendlyName6 rele6; WebButton6 rele6; FriendlyName7 rele7; WebButton7 rele7; FriendlyName8 rele8; WebButton8 rele8; Salo4; SwitchMode4 16; SwitchText5 stikalo5; SwitchMode5 16; SwitchText6 stikalo6; SwitchMode6 16;

Following:
Received response: 08:57:56.495 MQT: stat/small_tasmota32/RESULT = {"DeviceName":"ESP32 enota"}
Received response: 08:57:56.695 MQT: stat/small_tasmota32/RESULT = {"SetOption53":"ON"}
Received response: 08:57:56.897 MQT: stat/small_tasmota32/RESULT = {"FriendlyName1":"rele1"}
Received response: 08:57:57.099 MQT: stat/small_tasmota32/RESULT = {"WebButton1":"rele1"}
Received response: 08:57:57.301 MQT: stat/small_tasmota32/RESULT = {"FriendlyName2":"rele2"}
Received response: 08:57:57.504 MQT: stat/small_tasmota32/RESULT = {"WebButton2":"rele2"}
Received response: 08:57:57.705 MQT: stat/small_tasmota32/RESULT = {"FriendlyName3":"rele3"}
Received response: 08:57:57.907 MQT: stat/small_tasmota32/RESULT = {"WebButton3":"rele3"}
Received response: 08:57:58.160 MQT: stat/small_tasmota32/RESULT = {"FriendlyName4":"rele4"}
Received response: 08:57:58.412 MQT: stat/small_tasmota32/RESULT = {"WebButton4":"rele4"}
Received response: 08:57:58.664 MQT: stat/small_tasmota32/RESULT = {"FriendlyName5":"rele5"}
Received response: 08:57:58.915 MQT: stat/small_tasmota32/RESULT = {"WebButton5":"rele5"}
Received response: 08:57:59.117 MQT: stat/small_tasmota32/RESULT = {"FriendlyName6":"rele6"}
Received response: 08:57:59.369 MQT: stat/small_tasmota32/RESULT = {"WebButton6":"rele6"}
Received response: 08:57:59.622 MQT: stat/small_tasmota32/RESULT = {"FriendlyName7":"rele7"}
Received response: 08:57:59.822 MQT: stat/small_tasmota32/RESULT = {"WebButton7":"rele7"}
Received response: 08:58:00.074 MQT: stat/small_tasmota32/RESULT = {"FriendlyName8":"rele8"}
Received response: 08:58:00.275 MQT: stat/small_tasmota32/RESULT = {"WebButton8":"rele8"}
Received response: 08:58:00.533 MQT: stat/small_tasmota32/RESULT = {"Command":"Unknown"}
Received response: 08:58:00.731 MQT: stat/small_tasmota32/RESULT = {"SwitchMode4":16}
Received response: 08:58:00.931 MQT: stat/small_tasmota32/RESULT = {"SwitchText5":"stikalo5"}
Received response: 08:58:01.133 MQT: stat/small_tasmota32/RESULT = {"SwitchMode5":16}
Received response: 08:58:01.335 MQT: stat/small_tasmota32/RESULT = {"SwitchText6":"stikalo6"}
Received response: 08:58:01.536 MQT: stat/small_tasmota32/RESULT = {"SwitchMode6":16}

**Comment by msetina** [comment]
You can find Salo4; in the reply that is not in command. It is a garbled version of 
SwitchText1 stikalo1; SwitchMode1 16; SwitchText2 stikalo2; SwitchMode2 16; SwitchText3 stikalo3; SwitchMode3 16; SwitchText4 stikalo4;

**Reply by sfromis** [reply]
To me, that does not at all look like about the Backlog command, or a "length limit", but maybe about serial processing (you did not mention serial until now) getting into trouble keeping up with the pace of the long string of bytes. As a test, you could do the same command in the console, which I'd then expect to work.

**Reply by sfromis** [reply]
By the way. I just remembered that to do doing configs via serial, I made my serial setup slow down the "typing", instead of full speed 115200. Long time ago, don't remember details of what happened, but it works now.

**Reply by msetina** [reply]
OK. I thought serial would be the best/most stable way to load settings to device. If I want a new device I can use esptool to load tasmota to device, then I need to setup networking, which looks like the most straigthforward way to do is over serial. To be clean I was trying to do the template the same way. If setup was big I needed to limit the template command and now the Backlog to populate all the user oriented settings. If after setting the networking, HTTP would be a better way I have no problem with it. It is just a little cumbersom to switch the protocol. 
I know there is a web based tool, which I think is great, but it did not agree with my ESP32 board and its USB to serial implementation(maybe an omen). It worked with 8266 board from the same manufacturer.

**Reply by sfromis** [reply]
Well, serial **can** work, but if you are not able to do "slow typing", you may need to use shorter commands, and wait for each to complete after being sent, maybe just a fixed delay. And you need to support that Tasmota may do restarts, like after setting Module/Template, and also quite a few other settings, before being able to do certain other configs.

Unless using a binary with the setup built-in, what I'm doing currently has two stages. First I config the essentials via serial, obviously including network setup. This could be extended, but due to the restriction of needing restarts to do the full config, I have to split it. Therefore I do the rest of the config via MQTT, as I have a "bot" able to detect an unconfigured device coming online, and then I can publish commands with further details. This is also convenient when having the device reset for other reasons, allowing my automated config setup to work even when serial is not involved.

I prefer using MQTT over HTTP, as the latter does not facilitate that I can use a subscription to detect when the device is ready for more, but I do also have a variation allowing me to trigger the rest of the config with first command via HTTP, in case of me using the regular Tasmota onboarding to only config SSID/password. Still, I do find MQTT more convenient instead of messing with a serial command flow, and synchronize with each command being executed, and each restart needed.

**Reply by msetina** [reply]
I am using python to connect and "slow typing" is eluding me. I tried writing byte by byte with different sleeps between, but I am probably to high, because the other side does not respond at all.

**Reply by msetina** [reply]
Shuldn't there be RTS/CTS or XON/XOFF implemented on the other side to prevent this type of a problem?

**Reply by sfromis** [reply]
The word "should" does not apply. Arguably, it might be "nice to have", but that's a different thing. With Tasmota code aiming for being compact and usable, it does in general not try to be robust in edge cases.

I had no problem doing "slow typing" with a short delay between each byte. Not different from full speed code, except for the byte-by-byte loop.

**Reply by msetina** [reply]
OK. Wrong word. I agree, but would help

**Reply by sfromis** [reply]
The terminal program [tio](https://github.com/tio/tio) does support
`-o, --output-delay <ms>                Output character delay (default: 0)`
Not tested this, but with pipe input/output, looks like it might be usable by other means than manual.
